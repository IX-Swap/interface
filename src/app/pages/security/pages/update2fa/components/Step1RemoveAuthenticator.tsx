import React, { useEffect, useState } from 'react'
import { Typography, Grid, Button } from '@mui/material'
import { StepWrapper } from 'app/pages/security/components/StepWrapper'
import { useGetEmailCode } from 'app/pages/security/pages/update2fa/hooks/useGetEmailCode'
import { RemoveAuthenticatorForm } from 'app/pages/security/pages/update2fa/components/RemoveAuthenticatorForm'
import { useRemove2fa } from 'app/pages/security/pages/update2fa/hooks/useRemove2fa'
import { TwoFaData } from 'app/pages/security/types'
import { ResendCode } from 'app/pages/security/pages/update2fa/components/ResendCode/ResendCode'

export enum ScreenState {
  FirstScreen = 'Send code',
  SecondScreen = 'Remove and continue'
}

export interface RemoveCurrentAuthenticatorProps {
  onSuccessRemoveAuthenticator: (twoFaData: TwoFaData) => void
}

export const Step1RemoveAuthenticator = ({
  onSuccessRemoveAuthenticator
}: RemoveCurrentAuthenticatorProps) => {
  const [screenState, setScreenState] = useState<ScreenState>(
    ScreenState.FirstScreen
  )
  const { refetch, data, isLoading: isGetEmailLoading } = useGetEmailCode()
  const [update2fa, { isLoading: isUpdate2faLoading }] = useRemove2fa(
    onSuccessRemoveAuthenticator
  )
  const getEmailCode = async () => {
    await refetch()
  }

  useEffect(() => {
    if (data !== undefined) {
      setScreenState(ScreenState.SecondScreen)
    }
  }, [data])

  const renderSendCodeScreen = () => (
    <Grid container direction='column' spacing={3} alignItems='center'>
      <Grid item xs={12} md={8} lg={7}>
        <Typography align='center' variant='body1' style={{ maxWidth: 380 }}>
          To remove the Authenticator, we have to send a code to your email. The
          code will be valid for 30 minutes.
        </Typography>
      </Grid>
      <Grid item xs={12} md={8} lg={7}>
        <Button
          variant={'contained'}
          color={'primary'}
          disabled={isGetEmailLoading}
          onClick={getEmailCode}
        >
          Send code
        </Button>
      </Grid>
    </Grid>
  )

  const renderRemoveAndContinueScreen = () => {
    return (
      <Grid container direction={'column'} spacing={6}>
        <Grid item>
          <ResendCode action={getEmailCode} data={data} />
        </Grid>
        <Grid item>
          <RemoveAuthenticatorForm
            isLoading={isUpdate2faLoading}
            onSubmit={update2fa}
            email={data?.email ?? ''}
          />
        </Grid>
      </Grid>
    )
  }

  return (
    <StepWrapper title='Remove Authenticator by Verifying Your Identity'>
      {screenState === ScreenState.FirstScreen
        ? renderSendCodeScreen()
        : renderRemoveAndContinueScreen()}
    </StepWrapper>
  )
}
