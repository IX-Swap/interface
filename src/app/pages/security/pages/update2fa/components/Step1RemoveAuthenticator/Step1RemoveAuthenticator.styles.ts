import makeStyles from '@mui/styles/makeStyles'

export default makeStyles(theme => ({
  info: {
    margin: theme.spacing(5, 0, 7),
    width: '100%',
    maxWidth: '532px!important'
  },
  button: {
    width: '100%',
    maxWidth: 484,
    marginTop: theme.spacing(13.75)
  },
  formBlock: {
    maxWidth: '532px!important',
    width: '100%'
  },
  description: {
    marginTop: theme.spacing(2),
    marginBottom: theme.spacing(5),
    width: '100%',
    fontSize: 16,
    opacity: 0.8
  },
  icon: {
    width: '100%',
    maxWidth: 484
  }
}))
