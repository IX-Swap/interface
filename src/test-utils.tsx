import React from 'react'
import { Router } from 'react-router-dom'
import { render, RenderOptions, RenderResult } from '@testing-library/react'
import {
  createGenerateClassName,
  StylesProvider
} from '@material-ui/core/styles'
import { ThemeProvider } from '@material-ui/styles'
import Themes from './v2/themes'
import { history } from 'v2/history'
import { UserProvider } from 'v2/auth/context'
import { UserStore } from 'v2/auth/context/store'
import { PasswordResetProvider } from 'v2/auth/context/password-reset'
import PasswordResetStore from 'v2/auth/context/password-reset/store'
import { PasswordResetStep } from 'v2/auth/context/password-reset/types'
import { AuthorizerTableStore } from 'v2/app/pages/authorizer/context/store'
import { AuthorizerTableStoreProvider } from 'v2/app/pages/authorizer/context'
import { DepositStore } from 'v2/app/pages/accounts/pages/banks/context/store'
import { DepositStoreProvider } from 'v2/app/pages/accounts/pages/banks/context'
import { ServicesProvider } from 'v2/services/useServices'
import { renderHook, RenderHookResult } from '@testing-library/react-hooks'
import { BreadcrumbsProvider } from 'v2/hooks/useBreadcrumbs'

const generateClassName = createGenerateClassName({
  productionPrefix: 'ix'
})

const BaseProviders: React.FC = ({ children }) => {
  return (
    <StylesProvider generateClassName={generateClassName}>
      <ThemeProvider theme={Themes.default}>
        <BreadcrumbsProvider>
          <Router history={history}>{children}</Router>
        </BreadcrumbsProvider>
      </ThemeProvider>
    </StylesProvider>
  )
}

const customRenderer = (
  ui: any,
  options?: Omit<RenderOptions, 'queries'>
): RenderResult => {
  return render(ui, { wrapper: BaseProviders, ...options })
}

export const testComponentId = 'test-component'

export const fakeUserStore: Partial<UserStore> = {
  activeTab: 0,
  setActiveTab: jest.fn()
}

export const renderWithUserStore = (
  ui: any,
  store?: Partial<UserStore>
): RenderResult => {
  const WithUserProvider: React.FC = ({ children }) => (
    <BaseProviders>
      <UserProvider value={{ ...fakeUserStore, ...store }}>
        {children}
      </UserProvider>
    </BaseProviders>
  )

  return render(ui, { wrapper: WithUserProvider })
}

export const fakePasswordResetStore: Partial<PasswordResetStore> = {
  currentStep: PasswordResetStep.Request,
  setCurrentStep: jest.fn()
}

export const renderWithPasswordResetStore = (
  ui: any,
  store?: Partial<PasswordResetStore>
): RenderResult => {
  const WithUserProvider: React.FC = ({ children }) => (
    <BaseProviders>
      <PasswordResetProvider value={{ ...fakePasswordResetStore, ...store }}>
        {children}
      </PasswordResetProvider>
    </BaseProviders>
  )

  return render(ui, { wrapper: WithUserProvider })
}

export const fakeAuthorizerTableStore: Partial<AuthorizerTableStore> = {
  idKey: '__',
  uri: '/api/source',
  _getItemId: jest.fn()
}

export const renderWithAuthorizerTableStore = (
  ui: any,
  store?: Partial<AuthorizerTableStore>
): RenderResult => {
  const WithUserProvider: React.FC = ({ children }) => (
    <BaseProviders>
      <AuthorizerTableStoreProvider
        value={{ ...fakeAuthorizerTableStore, ...store }}
      >
        {children}
      </AuthorizerTableStoreProvider>
    </BaseProviders>
  )

  return render(ui, { wrapper: WithUserProvider })
}

export const fakeDepositStore: Partial<DepositStore> = {}

export const renderWithDepositStore = (
  ui: any,
  store?: Partial<DepositStore>
): RenderResult => {
  const WithUserProvider: React.FC = ({ children }) => (
    <BaseProviders>
      <DepositStoreProvider value={{ ...fakeDepositStore, ...store }}>
        {children}
      </DepositStoreProvider>
    </BaseProviders>
  )

  return render(ui, { wrapper: WithUserProvider })
}

export const renderHookWithServiceProvider = (
  hookFn: any,
  store: object = {}
): RenderHookResult<any, any> => {
  const WithServiceProvider: React.FC = ({ children }) => (
    <ServicesProvider value={store}>{children}</ServicesProvider>
  )

  return renderHook(hookFn, { wrapper: WithServiceProvider })
}

export * from '@testing-library/react'
export { customRenderer as render }
