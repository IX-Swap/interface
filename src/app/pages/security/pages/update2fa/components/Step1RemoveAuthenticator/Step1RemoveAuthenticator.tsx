import React from 'react'
import { Typography, Grid } from '@mui/material'
import { TwoFaData } from 'app/pages/security/types'
import { StepWrapper } from 'app/pages/security/components/StepWrapper'
import { useRemove2fa } from 'app/pages/security/pages/update2fa/hooks/useRemove2fa'
import { RemoveAuthenticatorForm } from 'app/pages/security/pages/update2fa/components/RemoveAuthenticatorForm'
import { VerificationInfo } from 'app/pages/security/pages/update2fa/components/VerificationInfo/VerificationInfo'
import useStyles from 'app/pages/security/pages/update2fa/components/Step1RemoveAuthenticator/Step1RemoveAuthenticator.styles'

export interface RemoveCurrentAuthenticatorProps {
  onSuccessRemoveAuthenticator: (twoFaData: TwoFaData) => void
}

export const Step1RemoveAuthenticator = ({
  onSuccessRemoveAuthenticator
}: RemoveCurrentAuthenticatorProps) => {
  const classes = useStyles()
  const [update2fa, { isLoading: isRemove2FALoading }] = useRemove2fa(
    onSuccessRemoveAuthenticator
  )

  return (
    <StepWrapper
      title={
        <>
          Update Authenticator by <br /> Verifying Your Identity
        </>
      }
    >
      <Grid container direction={'column'} alignItems={'center'}>
        <Grid item>
          <Typography
            align='center'
            variant='body1'
            className={classes.description}
          >
            To update your Authenticator please remove the current one
          </Typography>
        </Grid>
        <Grid item className={classes.info}>
          <VerificationInfo />
        </Grid>
        <Grid item className={classes.formBlock}>
          <RemoveAuthenticatorForm
            isRemove2FALoading={isRemove2FALoading}
            onSubmit={update2fa}
          />
        </Grid>
      </Grid>
    </StepWrapper>
  )
}
