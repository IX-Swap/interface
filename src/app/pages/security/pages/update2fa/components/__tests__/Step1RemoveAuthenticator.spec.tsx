import React from 'react'
import { render } from 'test-utils'
import { Step1RemoveAuthenticator } from 'app/pages/security/pages/update2fa/components/Step1RemoveAuthenticator/Step1RemoveAuthenticator'

describe('Step1RemoveAuthenticator', () => {
  const handleSuccessfulRemoveAuthenticator = jest.fn()

  afterEach(async () => {
    jest.clearAllMocks()
  })

  it('should match snapshot when useGetEmailCode has successful response', () => {
    const { container } = render(
      <Step1RemoveAuthenticator
        onSuccessRemoveAuthenticator={handleSuccessfulRemoveAuthenticator}
      />
    )

    expect(container).toMatchSnapshot()
  })
})
