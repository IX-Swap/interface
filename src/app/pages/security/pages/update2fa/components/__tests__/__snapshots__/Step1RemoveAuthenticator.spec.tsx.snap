// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Step1RemoveAuthenticator should match snapshot when useGetEmailCode has successful response 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column css-1vpwcmr-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
    >
      <h2
        class="MuiTypography-root MuiTypography-h2 MuiTypography-alignCenter css-2xfiep-MuiTypography-root"
      >
        Update Authenticator by 
        <br />
         Verifying Your Identity
      </h2>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column css-t0zib5-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 MuiTypography-alignCenter makeStyles-description-3 css-1gd5lqg-MuiTypography-root"
          >
            To update your Authenticator please remove the current one
          </p>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item makeStyles-info-1 css-13i4rnv-MuiGrid-root"
        >
          <div
            class="makeStyles-wrapper-4 MuiBox-root css-0"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 makeStyles-text-5 css-1b4l2pg-MuiTypography-root"
            >
              Verification code has been sent
            </p>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item makeStyles-formBlock-2 css-13i4rnv-MuiGrid-root"
        >
          <form
            data-testid="remove-authenticator-form"
            style="width: 100%;"
          >
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column css-1vpwcmr-MuiGrid-root"
            >
              <div
                class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-direction-xs-column css-80wlp8-MuiGrid-root"
              >
                <div
                  class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
                >
                  <div
                    class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root makeStyles-root-9 css-bxephj-MuiFormControl-root-MuiTextField-root"
                    control="[object Object]"
                    path="emailCode"
                  >
                    <label
                      class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-fhk4wr-MuiFormLabel-root-MuiInputLabel-root"
                      data-shrink="true"
                      for="emailCode"
                      id="emailCode-label"
                    >
                      Verification Code
                    </label>
                    <div
                      class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd css-1c4a1fp-MuiInputBase-root-MuiOutlinedInput-root"
                    >
                      <input
                        aria-invalid="false"
                        class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd css-1vvkud8-MuiInputBase-input-MuiOutlinedInput-input"
                        id="emailCode"
                        name="emailCode"
                        placeholder="E-mail verification code"
                        type="text"
                        value=""
                      />
                      <p
                        class="MuiTypography-root MuiTypography-body1 makeStyles-wrapper-12 css-1b4l2pg-MuiTypography-root"
                      >
                        Send
                      </p>
                      <fieldset
                        aria-hidden="true"
                        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                      >
                        <legend
                          class="css-173wfxe"
                        >
                          <span>
                            Verification Code
                          </span>
                        </legend>
                      </fieldset>
                    </div>
                  </div>
                </div>
                <div
                  class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body1 makeStyles-helperText-6 css-1b4l2pg-MuiTypography-root"
                  >
                    Enter the 6-digit code sent to 
                    
                  </p>
                </div>
              </div>
              <div
                class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-direction-xs-column css-80wlp8-MuiGrid-root"
              >
                <div
                  class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body1 makeStyles-label-7 css-1b4l2pg-MuiTypography-root"
                  >
                    Authenticator Code
                  </p>
                </div>
                <div
                  class="MuiGrid-root MuiGrid-container MuiGrid-item css-19hkhe-MuiGrid-root"
                >
                  <div
                    class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
                  >
                    <div
                      class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
                    >
                      <div
                        class="makeStyles-container-14 makeStyles-fullwidth-15"
                        style="display: flex;"
                      >
                        <div
                          style="display: flex; align-items: center;"
                        >
                          <input
                            aria-label="Please enter verification code. Digit 1"
                            autocomplete="off"
                            class="makeStyles-input-16  "
                            maxlength="1"
                            placeholder="_"
                            style="width: 1em; text-align: center;"
                            type="tel"
                            value=""
                          />
                        </div>
                        <div
                          style="display: flex; align-items: center;"
                        >
                          <input
                            aria-label="Digit 2"
                            autocomplete="off"
                            class="makeStyles-input-16 "
                            maxlength="1"
                            placeholder="_"
                            style="width: 1em; text-align: center;"
                            type="tel"
                            value=""
                          />
                        </div>
                        <div
                          style="display: flex; align-items: center;"
                        >
                          <input
                            aria-label="Digit 3"
                            autocomplete="off"
                            class="makeStyles-input-16 "
                            maxlength="1"
                            placeholder="_"
                            style="width: 1em; text-align: center;"
                            type="tel"
                            value=""
                          />
                        </div>
                        <div
                          style="display: flex; align-items: center;"
                        >
                          <input
                            aria-label="Digit 4"
                            autocomplete="off"
                            class="makeStyles-input-16 "
                            maxlength="1"
                            placeholder="_"
                            style="width: 1em; text-align: center;"
                            type="tel"
                            value=""
                          />
                        </div>
                        <div
                          style="display: flex; align-items: center;"
                        >
                          <input
                            aria-label="Digit 5"
                            autocomplete="off"
                            class="makeStyles-input-16 "
                            maxlength="1"
                            placeholder="_"
                            style="width: 1em; text-align: center;"
                            type="tel"
                            value=""
                          />
                        </div>
                        <div
                          style="display: flex; align-items: center;"
                        >
                          <input
                            aria-label="Digit 6"
                            autocomplete="off"
                            class="makeStyles-input-16 "
                            maxlength="1"
                            placeholder="_"
                            style="width: 1em; text-align: center;"
                            type="tel"
                            value=""
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body1 makeStyles-helperText-6 css-1b4l2pg-MuiTypography-root"
                  >
                    Enter the 6-digit code from your existing Authenticator App
                  </p>
                </div>
                <div
                  class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
                >
                  <button
                    class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-disableElevation MuiButtonBase-root Mui-disabled makeStyles-button-8 css-r7zh34-MuiButtonBase-root-MuiButton-root"
                    disabled=""
                    tabindex="-1"
                    type="submit"
                  >
                    Remove and Continue
                  </button>
                </div>
              </div>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</div>
`;
